### PR TITLE
Temp ffo

### PR DESCRIPTION
**Pull request by ghalym**
## Description
Created a new function block that extends FB_temp from lcls_twincat_general and added FFO capabilities. (Didn't add it to general because this meant I have to reference PMP in general, thoughts?)
The function block take a trip threshold as an input as well as a hysteresis and a veto boolean. The Function block then Faults based on the threshold and resets the fault based on the hysteresis percentage.
The ability to veto the device is added so the fault is not bypassed when the device is not in use or out of the beam.
The Veto logic of state to be evaluated in the user program.

## Motivation and Context
As per the SC PMP roadmap, this is a step into starting to integrate the temperature sensors into PMPS.

## How Has This Been Tested?
The function block was tested on a local PLC. A set of predefined test parameters where input while observing the output.

## Where Has This Been Documented?
In the function block.

## Pre-merge checklist
- [x] Libraries are set to ``Always Newest`` version (``Library, *``)
- [x] Code works interactively
- [x] Code contains descriptive comments
- [x] Test suite passes locally
- [ ] Committed with ``pre-commit`` or ran ``pre-commit run --all-files``
- [ ] Check that ``BP_IO`` parameters weren't modified unintentionally

### CODE REVIEW

File: lcls-twincat-pmps/PMPS/MajorComponents/FB_TempSensor_FFO.TcPOU [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+bConnected := NOT (bOverrange AND bError);
+fTemp := INT_TO_LREAL(iRaw) * fResolution;
+
+//Verify Hysteresis is between 1-100, Shouldn't be 0. shouldn't be a 100 either.
+fHysteresis:=  LIMIT(1, fHysteresis, 100);
+
+// Evaluate the threshold trip point
+IF (fTemp >= fFaultThreshold) OR NOT bConnected THEN
+	bFAULT_OK := FALSE;
+ELSIF (fTemp < (fFaultThreshold - fFaultThreshold*fHysteresis/100)) AND bConnected THEN
```
Comment (klauer) [inline lcls-twincat-pmps/PMPS/MajorComponents/FB_TempSensor_FFO.TcPOU:53]:
```suggestion
ELSIF fTemp < (fFaultThreshold - fFaultThreshold*fHysteresis/100) THEN
```

The `NOT bConnected` is caught in the first clause, which means `bConnected` has to be True here.

File: lcls-twincat-pmps/PMPS/MajorComponents/FB_TempSensor_FFO.TcPOU [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	bFAULT_OK := TRUE;
+END_IF
```
Comment (klauer) [inline lcls-twincat-pmps/PMPS/MajorComponents/FB_TempSensor_FFO.TcPOU:55]: Thinking about the values here, the implicit `ELSE` clause would be:
`(fFaultThreshold - fFaultThreshold * fHysteresis / 100) < fTemp < fFaultThreshold`
The area in which you want to just keep `bFAULT_OK` as the previous value, correct?

File: lcls-twincat-pmps/PMPS/MajorComponents/FB_TempSensor_FFO.TcPOU [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	rtRESET	:	R_TRIG;
+	ftFAULT	:	F_TRIG;
+	ftConnected	:	F_TRIG;
+	fbLogger : FB_LogMessage := (eSubsystem:=E_SubSystem.MPS, nMinTimeViolationAcceptable:=10);
+END_VAR
+]]></Declaration>
+    <Implementation>
+      <ST><![CDATA[ // The manual states that we are disconnected if we are both overrange and in an error state
+bConnected := NOT (bOverrange AND bError);
+fTemp := INT_TO_LREAL(iRaw) * fResolution;
```
Comment (ZLLentz) [inline lcls-twincat-pmps/PMPS/MajorComponents/FB_TempSensor_FFO.TcPOU:45]: Reviewer note: above two lines copied from upstream FB in general lib

**Review by klauer — APPROVED** [review]
The logic and implementation seem reasonable to me.

As to the PR question: it's sensible to keep general-but-PMPS-specific function blocks in lcls-twincat-pmps, in my opinion

File: lcls-twincat-pmps/PMPS/MajorComponents/FB_TempSensor_FFO.TcPOU [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+//Verify Hysteresis is between 1-100, Shouldn't be 0. shouldn't be a 100 either.
+fHysteresis:=  LIMIT(1, fHysteresis, 100);
```
Comment (ZLLentz) [inline lcls-twincat-pmps/PMPS/MajorComponents/FB_TempSensor_FFO.TcPOU:48]: I think 1 is a good choice for a lower limit. Imagine a typical threshold of like 50C -> 1% is 49.5C

File: lcls-twincat-pmps/PMPS/MajorComponents/FB_TempSensor_FFO.TcPOU (hunk 2 above)
Comment (ZLLentz) [inline lcls-twincat-pmps/PMPS/MajorComponents/FB_TempSensor_FFO.TcPOU:55]: You don't have to change it but maybe it's more readable to split up the logic
```suggestion
IF fTemp >= fFaultThreshold THEN
	bFAULT_OK := FALSE;
ELSIF fTemp < (fFaultThreshold - fFaultThreshold*fHysteresis/100) THEN
	bFAULT_OK := TRUE;
END_IF
bFAULT_OK R= NOT bConnected
```

File: lcls-twincat-pmps/PMPS/MajorComponents/FB_TempSensor_FFO.TcPOU [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+END_IF
+
+ACT_Logger();
+
+(*FAST FAULT*)
+FFO(i_xOK := bFAULT_OK OR bVeto,
+    i_xReset :=,
+	i_xAutoReset :=,
+	i_DevName := sDevName,
+    io_fbFFHWO := io_fbFFHWO);]]></ST>
```
Comment (ZLLentz) [inline lcls-twincat-pmps/PMPS/MajorComponents/FB_TempSensor_FFO.TcPOU:64]: In your comment at the top:
```
	bVeto : BOOL:=FALSE; // This Fault can be Vetoed if the bVeto is TRUE
```
So did you mean:
```suggestion
FFO(i_xOK := bFAULT_OK,
    i_xReset :=,
    i_xAutoReset :=,
    i_xVetoable := bVeto,
    i_DevName := sDevName,
    io_fbFFHWO := io_fbFFHWO);]]></ST>
```

**Review by ZLLentz — COMMENTED** [review]
I caught one thing that looks off with the veto, the rest is a mix of good job :+1: and optionals

File: lcls-twincat-pmps/PMPS/MajorComponents/FB_TempSensor_FFO.TcPOU (hunk 5 above)
Comment (ghalym) [reply lcls-twincat-pmps/PMPS/MajorComponents/FB_TempSensor_FFO.TcPOU:64]: Maybe I need to change the comment to say that this should not fault if the bVeto is true. Then the original implementation stays i_xOK := bFAULT_OK OR bVeto.

**Comment by ghalym** [comment]
I also need to fix some checks that failed.

File: lcls-twincat-pmps/PMPS/MajorComponents/FB_TempSensor_FFO.TcPOU (hunk 5 above)
Comment (ZLLentz) [reply lcls-twincat-pmps/PMPS/MajorComponents/FB_TempSensor_FFO.TcPOU:64]: Yes, changing the comment is also fine :+1:, I just want to make sure we're implementing exactly what we intend to + document it appropriately.

File: lcls-twincat-pmps/PMPS/MajorComponents/FB_TempSensor_FFO.TcPOU (hunk 2 above)
Comment (ghalym) [reply lcls-twincat-pmps/PMPS/MajorComponents/FB_TempSensor_FFO.TcPOU:55]: yes.

**Review by ZLLentz — APPROVED** [review]
I missed the re-review request- good catch on the pytmc syntax!

**Pull request merged**